maintenance/annotation_maint.py

```diff
@@ -1,38 +1,147 @@
+            assert merged_intervals == merged(merged_intervals, closed=closed)
+
+    assert merged([(10, 20), (21, 30)], closed=True) == [(10, 30)]
+    assert merged([(10, 20), (21, 30)], closed=False) == [(10, 20), (21, 30)]
+
+
+def gff_recarray_to_stdpopsim_intervals(gff):
+    """
+    Merge overlapping intervals and convert coordinates. GFF intervals are
+    1-based [i,j], but stdpopsim intervals are 0-based [i-1,j).
+    """
+    intervals = np.array(merged(zip(gff.start, gff.end), closed=True))
+    intervals[:, 0] = intervals[:, 0] - 1
+    return intervals
+
+
+def get_gff_recarray(url, sha256):
+    local_path = pathlib.Path(url).name
+
+    if not pathlib.Path(local_path).exists():
+        print(f"downloading {url}")
+        stdpopsim.utils.download(url, local_path)
+
+    print("checking sha256")
+    local_sha256 = stdpopsim.utils.sha256(local_path)
+    if local_sha256 != sha256:
+        print(
+            f"{local_path}: sha256: expected {sha256}, but found {local_sha256}. "
+            "Delete the file to download it again."
+        )
+        exit(1)
+
+    print(f"loading {local_path} into numpy recarray")
+    gff = allel.gff3_to_recarray(local_path)
+    return gff
+
+
+def make_tarfile(output_filename, source_dir, dest):
+    with tarfile.open(output_filename, "w:gz") as tar:
+        tar.add(source_dir, arcname=dest)
+    tar.close()
+
+
 # loop through species and download
-for spc in stp.all_species():
+for spc in stdpopsim.all_species():
     if spc.annotations:
-        address = spc.annotations[0].url
-        genome_version = os.path.basename(address).split(".")[1]
-        logger.info(f"Downloading GFF file {spc.id}")
-        tmp_path = f"{spc.id}.tmp.gff.gz"
-        try:
-            x, y = urllib.request.urlretrieve(address, tmp_path)
-        except FileNotFoundError:
-            warnings.warn("can't connnect to url")
-        logger.info(f"creating zarr arrays {spc.id}")
-        # create zarr store and zarr root
-        spc_path = os.path.join(annot_path, spc.id + "." + genome_version + ".zip")
-        store = zarr.ZipStore(spc_path)
-        root = zarr.group(store=store, overwrite=True)
-        x = allel.gff3_to_dataframe(tmp_path)
-        for col_name in x.columns:
-            if x[col_name].dtype == "O":
-                tmp = root.array(col_name, np.array(x[col_name], dtype=str))
-            else:
-                tmp = root.array(col_name, np.array(x[col_name]))
-        # cleanup
-        os.unlink(tmp_path)
-        store.close()
+        for an in spc.annotations:
+            GFF_URL = an.url
+            GFF_SHA256 = an.gff_sha256
+            CHROM_IDS = [chrom.id for chrom in spc.genome.chromosomes]
+            genome_version = os.path.basename(GFF_URL).split(".")[1]
+            logger.info(f"Downloading GFF file {spc.id}")
+            tmp_path = f"{spc.id}.tmp.gff.gz"
+            gff = get_gff_recarray(GFF_URL, GFF_SHA256)
```

Comment:
I think the UPPERCASE variables aren't really needed here. Just replace the use of these variables with `an.url` and `an.gff_sha256`, etc.